stdlib/public/SDK/Foundation/Data.swift

```diff
@@ -24,54 +24,773 @@ internal func __NSDataInvokeDeallocatorFree(_ mem: UnsafeMutableRawPointer, _ le
+        _backing = .immutable(immutableReference)
+    }
+    
+    public init(mutableReference: NSMutableData) {
+        _bytes = mutableReference.mutableBytes
+        _capacity = 0
+        _needToZero = false
+        _length = mutableReference.length
+        _backing = .mutable(mutableReference)
+    }
+    
+    public init(customReference: NSData) {
+        _bytes = nil
+        _capacity = 0
+        _needToZero = false
+        _length = 0
+        _backing = .customReference(customReference)
+    }
+    
+    public init(customMutableReference: NSMutableData) {
+        _bytes = nil
+        _capacity = 0
+        _needToZero = false
+        _length = 0
+        _backing = .customMutableReference(customMutableReference)
     }
     
     deinit {
-        releaseWrappedObject()
+        switch _backing {
+        case .swift:
+            _freeBytes()
+            break
+        default:
+            break
+        }
+    }
+    
+    @inline(__always)
+    public func mutableCopy() -> _DataStorage {
+        switch _backing {
+        case .swift:
+            return _DataStorage(bytes: _bytes, length: _length, copy: true, deallocator: nil)
+        case .immutable(let d):
+            return _DataStorage(mutableReference: d.mutableCopy() as! NSMutableData)
+        case .mutable(let d):
+            return _DataStorage(mutableReference: d.mutableCopy() as! NSMutableData)
+        case .customReference(let d):
+            return _DataStorage(customMutableReference: d.mutableCopy() as! NSMutableData)
+        case .customMutableReference(let d):
+            return _DataStorage(customMutableReference: d.mutableCopy() as! NSMutableData)
+        }
+    }
+    
+    public func withInteriorPointerReference<T>(_ work: (NSData) throws -> T) rethrows -> T {
+        switch _backing {
+        case .swift:
+            return try work(NSData(bytesNoCopy: _bytes!, length: _length, freeWhenDone: false))
+        case .immutable(let d):
+            return try work(d)
+        case .mutable(let d):
+            return try work(d)
+        case .customReference(let d):
+            return try work(d)
+        case .customMutableReference(let d):
+            return try work(d)
+        }
+    }
+    
+    public func bridgedReference() -> NSData {
+        switch _backing {
+        case .swift:
+            return _NSSwiftData(backing: self)
+        case .immutable(let d):
+            return d
+        case .mutable(let d):
+            return d
+        case .customReference(let d):
+            return d
+        case .customMutableReference(let d):
+            // Because this is returning an object that may be mutated in the future it needs to create a copy to prevent
+            // any further mutations out from under the reciever
+            return d.copy() as! NSData
+        }
+    }
+    
+    public static func ==(_ lhs: _DataStorage, _ rhs: _DataStorage) -> Bool {
+        switch (lhs._backing, rhs._backing) {
```

Comment:
I was attempting to avoid blocks since I was seeing a ton of extra ref counting even in whole module optimizations but this explosion didn't have those.